devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -0,0 +1,277 @@
+
+## Key Benefit Goals
+
+1. **Separation of Concerns**: Decision-making is isolated from execution
+2. **Immutable Data**: All data structures are immutable, preventing side effects
+3. **Testability**: Each phase can be tested independently with predictable inputs/outputs
+4. **Maintainability**: Clear boundaries between reading, processing, building, and writing
+5. **Flexibility**: New readers, processors, or outputs can be added without affecting other phases
+6. **Error Handling**: Failures are contained within phases and can be handled appropriately
+
+## Success Criteria
+
+- [ ] All I/O operations occur only at pipeline boundaries
+- [ ] Decision-making is separated from execution
+- [ ] All data structures are immutable
+- [ ] Each pipeline phase can be tested independently
+- [ ] Existing functionality is preserved
+- [ ] Performance is maintained or improved
+- [ ] Code coverage remains above 90%
+
+## Risk Mitigation
+
+1. **Feature Flags**: Toggle between old/new implementations -- either through CLI or .INI
+2. **Gradual Migration**: One component / feature at a time
+3. **Comprehensive Testing**: Test each change thoroughly
+4. **Performance Monitoring**: Track pipeline performance throughout migration
+
+This refactoring plan ensures the pipeline becomes more maintainable, testable, and extensible while minimizing risk through incremental changes that preserve existing functionality.
+
+## New Pipeline Architecture
+
+### Pipeline Diagram
+
+```
+┌─────────────────────────────────────────────────────────────────┐
+│                    DECISION PHASE                               │
+├─────────────────────────────────────────────────────────────────┤
+│ 1. Configuration Analysis                                       │
+│    └─ Parse config → Determine processing strategy              │
+│                                                                 │
+│ 2. File Discovery & Grouping                                   │
+│    └─ Scan directories → Group related files                   │
+│                                                                 │
+│ 3. Reader Selection                                             │
+│    └─ Analyze file types → Select appropriate readers          │
+│                                                                 │
+│ 4. Geometry Strategy                                            │
+│    └─ Analyze spatial config → Choose geometry approach        │
+│                                                                 │
+│ 5. Template Selection                                           │
+│    └─ Based on outputs → Select UMM-G/CNM templates           │
+└─────────────────────────────────────────────────────────────────┘
+                                │
+                                ▼
+┌─────────────────────────────────────────────────────────────────┐
+│                     READING PHASE                              │
+├─────────────────────────────────────────────────────────────────┤
```

Comment:
So far, we've needed to do some reading before any processing strategy can be determined.

Reply:
Yeah, this isn't right. I'll rework the diagram.